draft-ietf-radext-radiusdtls-bis.md

```diff
@@ -236,31 +236,31 @@ RADIUS/(D)TLS client implementations SHOULD implement this model, but MUST imple
-  * It is possible for a RADIUS/(D)TLS server to not require additional name checks for incoming RADIUS/(D)TLS clients, i.e. if the client used dynamic lookup.
-    In this case, the certificate is accepted immediately after the {{RFC5280}} trust chain checks.
-    This MUST NOT be used outside of trusted network environments or without additional certificate attribute checks in place.
-* Implementations MAY allow a configuration of a set of additional properties of the certificate to check for a peer's authorization to communicate (e.g. a set of allowed values in subjectAltName:URI or a set of allowed X.509v3 Certificate Policies).
-* When the configured trust base changes (e.g., removal of a CA from the list of trusted CAs; issuance of a new CRL for a given CA), implementations SHOULD renegotiate the TLS session to reassess the connecting peer's continued authorization.[^may-should-trustbase]{:jf}
+* When the configured trust base changes (e.g., removal of a CA from the list of trusted CAs; issuance of a new CRL for a given CA), implementations SHOULD reassess all connected peer's continued authorization. Note that TLS 1.3 no longer supports renegotiation to fulfill this requirement. [^may-should-trustbase]{:jf}
 
 [^may-should-trustbase]: Open discussion: RFC6614 says "may" here. I think this should be a "should". There are some discussions to change this to "must". Input from TLS/UTA experts is appreciated.
 
+RADIUS/(D)TLS clients and server MUST follow {{!RFC9525}} when validating peer identities. Specific details are provided below:
+* The Common Name RDN MUST NOT be used to identify peers
+* Certifcates MAY use wildcards in the identifiers of DNS names and realm names, but only as the complete, left-most label.
+* RADIUS/(D)TLS clients validate the servers identity to match their local configuration, accepting the identity on the first match:
+  - If the expected RADIUS/(D)TLS server is associated with a specific NAI realm, e.g. by dynamic discovery {{!RFC7585}} or static configuration, that realm is matched against the presented realm names in the subjectAltName:naiRealm extension.
+  - If the expected RADIUS/(D)TLS server was configured as a hostname, or yielded by a dynamic discovery procedure, the configured name is matched against the presented names from the subjectAltName:DNS extension. Since a dynamic discovery might by itself not be secured, implementations MAY require the use of DNSSEC {{!RFC4033}} to ensure the authenticity of the DNS result before considering this identity as valid. 
+  - If the expected RADIUS/(D)TLS server was configured as an IP address, the configured IP address is matched against the presented addresses in the subjectAltName:iPAddr extension.
+  - Clients MAY use other attributes of the certificate to validate the servers identity, but it MUST NOT connect to servers without doing so. Omitting the identity verification can allow attackers to loop RADIUS traffic back to the originating client.
```

Comment:
I would be more explicit here, because the "without doing so" is a bit vague.
What is it exactly that we mean? Is it sufficient that the client checks the "issuer" attribute of the certificate?
Why is this sentence of looping just here? The loop attack could possibly also happen with other scenarios (like DNS without DNSSEC and the NAPTR just points back to the server)

Reply:
The intention here is that a client MUST identy the server by any element that clearly identifies it. I think this includes subject, subjectAltNames, issuer+certificate serial (but not the issuer by itself).

The loop attack refers to the same issue we discussed regarding TLS-PSK, that an attacker could loop the TLS packets (without decrypting them) back to the client and it would accept that connection (if it generally accepts incomming connections, i.e. proxies).

Reply:
agreed to both points. This was as a last minute addition anyway, after it got mentioned in the TLS-PSK context.

Reply:
>  Second, the Loop attack is a general attack, so maybe it is worth to put it in security considerations, have a few words here about it and then reference the security considerations section for details?

One workaround would be to check that the client/server random (in OpenSSLese, `SSL_get_client_random` and `SSL_get_server_random`) do not match.

Reply:
don't know how exactly client/server random works, but even experimentally, I'm struggling to see how this would work.

If I call `get_client_random` and `get_server_random` both on the connecting and accepting side of a tls connection, I actually get 4 different values, even when both sides run in the same process and use the same ssl context.

Am I missing something?

Reply:
> don't know how exactly client/server random works, but even experimentally, I'm struggling to see how this would work.
> 
> If I call `get_client_random` and `get_server_random` both on the connecting and accepting side of a tls connection, I actually get 4 different values, even when both sides run in the same process and use the same ssl context.
> 
> Am I missing something?

The first four bytes is the `gmt_unix_time` (legacy, and I think now purely random) when shown in `openssl s_client ...`, so that may be it, otherwise Works for Me(tm).

Apply the ghetto patch to https://github.com/openssl/openssl/tree/master/demos/sslecho, compile and run `sslecho s`:
```
diff --git a/demos/sslecho/main.c b/demos/sslecho/main.c
index eec8e13a47..b4fca946cd 100644
--- a/demos/sslecho/main.c
+++ b/demos/sslecho/main.c
@@ -7,6 +7,7 @@
  *  https://www.openssl.org/source/license.html
  */
 
+#include <assert.h>
 #include <stdio.h>
 #include <unistd.h>
 #include <string.h>
@@ -213,6 +214,18 @@ int main(int argc, char **argv)
 
                 printf("Client SSL connection accepted\n\n");
 
+                unsigned char random[32];
+                assert(SSL_get_client_random(ssl, random, sizeof(random)) == sizeof(random));
+                printf("client random: ");
+                for (int i = 0; i < sizeof(random); i++)
+                    printf("%02X", random[i]);
+                printf("\n");
+                assert(SSL_get_server_random(ssl, random, sizeof(random)) == sizeof(random));
+                printf("server random: ");
+                for (int i = 0; i < sizeof(random); i++)
+                    printf("%02X", random[i]);
+                printf("\n");
+
                 /* Echo loop */
                 while (true) {
                     /* Get message from client; will fail if client closes connection */
```

As the client run `openssl s_client -trace -connect localhost:4433`:
```
CONNECTED(00000003)
Sent Record
Header:
  Version = TLS 1.0 (0x301)
  Content Type = Handshake (22)
  Length = 292
    ClientHello, Length=288
      client_version=0x303 (TLS 1.2)
      Random:
        gmt_unix_time=0x919DE39C
        random_bytes (len=28): 97E42A6040AD6B9FA248D29B5C2C153E3F92AD9E14412CB875F8BF78
      session_id (len=32): 1557C0A035FB2A3DF97F5067FC534FC9E71BC106FB929E823A998467C893C6A0
      cipher_suites (len=62)

[snipped]

Received Record
Header:
  Version = TLS 1.2 (0x303)
  Content Type = Handshake (22)
  Length = 122
    ServerHello, Length=118
      server_version=0x303 (TLS 1.2)
      Random:
        gmt_unix_time=0x99B5F850
        random_bytes (len=28): 30BCD49D1F53AFE61C5A63B16AD7FABE3B444EEE5D1093D78E878639
      session_id (len=32): 1557C0A035FB2A3DF97F5067FC534FC9E71BC106FB929E823A998467C893C6A0
      cipher_suite {0x13, 0x02} TLS_AES_256_GCM_SHA384
      compression_method: No Compression (0x00)

[snipped]
```

The output I got matched.

Reply:
Thanks for the detailed explanation; looks like I goofed up my experiment.

So yes, one could compare client/server randoms, but upon opening (or accepting) a connection, you would have to compare its random to verey other TLS connection. I.e. the client and server random of a single conneciton are always different (I guess as they should be for security reasons)

Reply:
I think you only need to limit the check window to the pool of the systems own not-yet-established TLS connections, as those connections have to come from yourself and be active for this attack to work, right?

Reply:
Yes, but still this might be tricky to implement for the exact timing and potential race-conditions. After all the client and server side code very intentionally run at the same time.

Reply:
During the client hello callback (`SSL_CTX_set_client_hello_cb`) the connection is not established and the server side is yet to fully process the request.

Maybe I am missing something, I do not see a race. If there was, you would be unable to use this function to select a server certificate (ie. ALPN) which is one of the primary use cases for it.

It does become slightly more complicated if the 'server' is actually a pool of nodes, but all that would require is an RPC call out to the other nodes to see what connections are outstanding.

Meanwhile the connection under inspection is frozen in time until the callback completes.

Can you think of a situation where this fails?

I am only lobbying for it, as if this is a good defence (which implementing aside we still do not know) it is trivial to implement. Helping us is the connection pool churn is low and so the length of the list of being established in flight TLS connections is low (ie. less than 1000).

Reply:
Its two threads (or more, at least for me) and a network connection, which for me always puts on the warning light, be careful. There can always be events from the network (timeouts, a connecting being closed) and the race conditions may arise when two legitimate connections happen a the same time. In the end one thread has to access data from another, which requires careful coordination.

I do agree that it's a good idea and definitely worth a shot, I just anticipate that its not going to be as trivial to implement as at first glance it might seem.

Besides that, these are implementaiton details. I guess the question we should ask here is what we should recommend or even mandate in the RFC.

Reply:
> Its two threads (or more, at least for me) and a network connection, which for me always puts on the warning light, be careful.

Sure, but you have described the effect of language and design decisions and the usual trade off and balances we all make. For others, sharing memory may not even be possible in their language (functional, actor model, etc) so this would simply not be a concern for them.

As such, definitely I would agree that mandating this would be unfair to implementers but it would be well placed as recommendations on one of several ways to avoid this problem.

We do need something though as related to this all is https://github.com/radext-wg/draft-ietf-radext-tls-psk/issues/4 where stated is a KDF should be used (and crucially 'context' populated) but the two major public implementations do not.

If the two-step KDF 'dance' is to be dropped, something would be needed and maybe checking handshake random numbers and provide the backwards compatible sticky tape.